config/mysqlpersister/mysqlpersister.go

```diff
@@ -0,0 +1,210 @@
+package mysqlpersister
+
+import (
+	"errors"
+	"fmt"
+	"sort"
+	"sync"
+	"time"
+
+	sq "github.com/Masterminds/squirrel"
+	_ "github.com/go-sql-driver/mysql"
+	"github.com/golang/protobuf/proto"
+	"github.com/jmoiron/sqlx"
+
+	"github.com/square/quotaservice/config"
+	"github.com/square/quotaservice/logging"
+	qsc "github.com/square/quotaservice/protos/config"
+)
+
+type mysqlPersister struct {
+	latestVersion int
+	db            *sqlx.DB
+	m             *sync.RWMutex
+
+	watcher  chan struct{}
+	shutdown chan struct{}
+
+	activeFetchers *sync.WaitGroup
+
+	configs map[int]*qsc.ServiceConfig
+}
+
+type configRow struct {
+	Version int    `db:"Version"`
+	Config  string `db:"Config"`
+}
+
+func New(dbUser, dbPass, dbHost string, dbPort int, dbName string, pollingInterval time.Duration) (config.ConfigPersister, error) {
+	db, err := sqlx.Open("mysql",
+		fmt.Sprintf("%s:%s@(%s:%v)/%s",
+			dbUser,
+			dbPass,
+			dbHost,
+			dbPort,
+			dbName))
+
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = db.Query("SELECT 1 FROM quotaservice LIMIT 1")
+	if err != nil {
+		return nil, errors.New("table quotaservice does not exist")
+	}
+
+	mp := &mysqlPersister{
+		db:             db,
+		configs:        make(map[int]*qsc.ServiceConfig),
+		activeFetchers: &sync.WaitGroup{},
+		m:              &sync.RWMutex{},
+		watcher:        make(chan struct{}),
+		shutdown:       make(chan struct{}),
+		latestVersion:  -1,
+	}
+
+	mp.pullConfigs()
+
+	mp.activeFetchers.Add(1)
+	go mp.configFetcher(pollingInterval)
+
+	return mp, nil
+}
+
+func (mp *mysqlPersister) configFetcher(pollingInterval time.Duration) {
+	defer mp.activeFetchers.Done()
+
+	for {
+		select {
+		case <-time.After(pollingInterval):
+			if mp.pullConfigs() {
+				mp.notifyWatcher()
+			}
+		case <-mp.shutdown:
+			logging.Print("Received shutdown signal, shutting down mysql watcher")
+			return
+		}
+	}
+}
+
+// pullConfigs checks the database for new configs and returns true if there is a new config
+func (mp *mysqlPersister) pullConfigs() bool {
+	mp.m.RLock()
+	v := mp.latestVersion
+	mp.m.RUnlock()
+
+	q, args, err := sq.
+		Select("Version", "Config").
+		From("quotaservice").
+		Where("Version > ?", v).
+		OrderBy("Version ASC").ToSql()
+	if err != nil {
+		logging.Printf("Could not generate query to fetch config updates: %s", err)
+		return false
+	}
+
+	var rows []*configRow
+	err = mp.db.Select(&rows, q, args...)
+	if err != nil {
+		logging.Printf("Received error from querying mysql for the latest configs mysql: %s", err)
```

Comment:
If we bubbled this error up, do you think it would be more clear to debuggers (especially those that aren't reviewing the code now) where the error is coming from?

Reply:
I can't think of a way to bubble this up?  We could feasibly pass it to the `GetPersistedConfig` but I don't think we want to induce errors in that if the mysql server is down and we have configs cached.  I'm open to other places we could bubble this up to?

This behaviour is analogous to the way the zookeeper persister handles errors fetching configs.

Reply:
Bubbling may have been a weird phrase. I meant returning the error here and logging in a function further up the call stack. E.g., configFetcher() would handle the error.

By returning errors, the caller can wrap it to add additional context. If pullConfigs() is only called in one place, this may not be a big deal.

Reply:
Forgot that it's called in multiple places. Bubbled this up